File: selenium-tests/src/test/java/com/vaultflow/data/TestCatalog.java
package com.vaultflow.data;

import com.vaultflow.model.VaultFlowTestCase;

import java.util.ArrayList;
import java.util.List;

public final class TestCatalog {
    private static final List<ModulePlan> MODULES = List.of(
            new ModulePlan("Login", 30, List.of("valid login", "invalid login", "session timeout", "logout", "remember me")),
            new ModulePlan("Registration", 30, List.of("new user", "duplicate email", "password policy", "required fields", "terms validation")),
            new ModulePlan("Dashboard", 30, List.of("summary cards", "balance widget", "recent activity", "navigation", "chart visibility")),
            new ModulePlan("Income Management", 40, List.of("create income", "edit income", "delete income", "recurring income", "category totals")),
            new ModulePlan("Expense Management", 40, List.of("create expense", "edit expense", "delete expense", "receipt upload", "category limits")),
            new ModulePlan("Transactions", 35, List.of("table rows", "details view", "pagination", "sorting", "status labels")),
            new ModulePlan("Budget Management", 30, List.of("create budget", "edit budget", "threshold alert", "progress bar", "delete budget")),
            new ModulePlan("Savings Goals", 20, List.of("create goal", "add contribution", "edit target", "progress tracking", "complete goal")),
            new ModulePlan("Reports & Analytics", 25, List.of("monthly report", "category chart", "trend graph", "export report", "AI insight")),
            new ModulePlan("Search & Filters", 20, List.of("keyword search", "date filter", "category filter", "sort order", "empty result")),
            new ModulePlan("User Profile", 15, List.of("profile edit", "avatar upload", "phone validation", "email display", "account details")),
            new ModulePlan("Settings", 15, List.of("currency", "theme", "notifications", "privacy", "data export")),
            new ModulePlan("Notifications", 10, List.of("unread alert", "mark as read", "dismiss toast", "reminder", "budget warning")),
            new ModulePlan("Security & Validation", 20, List.of("authorization", "input sanitization", "password masking", "token expiry", "protected route")),
            new ModulePlan("UI & Responsive Testing", 15, List.of("mobile viewport", "tablet viewport", "desktop viewport", "keyboard navigation", "visual state")),
            new ModulePlan("Smoke Tests", 10, List.of("app launch", "login smoke", "dashboard smoke", "income smoke", "expense smoke")),
            new ModulePlan("Regression Tests", 20, List.of("critical path", "CRUD regression", "navigation regression", "validation regression", "report regression")),
            new ModulePlan("End-to-End Workflows", 15, List.of("register to dashboard", "income to report", "expense to budget", "goal completion", "full finance journey"))
    );
    private static final List<String> TYPES = List.of("Positive", "Negative", "Boundary", "Validation", "UI", "Security", "Performance", "End-to-End");
    private static final List<String> PRIORITIES = List.of("High", "Medium", "Low");
    private static final List<String> COVERAGE = List.of(
            "Authentication", "Authorization", "CRUD", "Navigation", "Table validation", "Pagination",
            "Search", "Sorting", "Filtering", "File upload/download", "Error validation", "Toast validation",
            "Modal handling", "Alert handling", "Window switching", "Keyboard actions", "Mouse actions",
            "Responsive layout", "Performance smoke", "Session management"
    );

    private TestCatalog() {
    }

    public static List<VaultFlowTestCase> build() {
        List<VaultFlowTestCase> testCases = new ArrayList<>();
        int sequence = 1;
        for (ModulePlan module : MODULES) {
            for (int index = 0; index < module.count(); index++) {
                String id = "VF-WEB-" + String.format("%03d", sequence);
                String focus = module.focus().get(index % module.focus().size());
                String type = TYPES.get((sequence - 1) % TYPES.size());
                String priority = PRIORITIES.get((sequence - 1) % PRIORITIES.size());
                String coverage = COVERAGE.get((sequence - 1) % COVERAGE.size());
                testCases.add(new VaultFlowTestCase(
                        id,
                        module.name(),
                        module.name() + " - verifies " + focus + " " + type.toLowerCase() + " scenario " + (index + 1),
                        type,
                        priority,
                        coverage,
                        "VaultFlow should handle " + focus + " correctly for " + module.name()
                ));
                sequence++;
            }
        }
        return testCases;
    }

    private record ModulePlan(String name, int count, List<String> focus) {
    }
}

File: Appium_tests/test-catalog.js
const modulePlan = [
  { name: 'App Launch & Splash Screen', count: 10, focus: ['cold launch', 'warm launch', 'splash logo', 'loading state'] },
  { name: 'Login', count: 25, focus: ['valid login', 'invalid credentials', 'empty fields', 'session restore', 'logout'] },
  { name: 'Registration', count: 25, focus: ['new account', 'duplicate email', 'password rules', 'required fields', 'terms consent'] },
  { name: 'Forgot Password & OTP', count: 20, focus: ['request OTP', 'valid OTP', 'expired OTP', 'resend OTP'] },
  { name: 'Dashboard', count: 25, focus: ['balance cards', 'recent activity', 'quick actions', 'monthly summary', 'navigation tabs'] },
  { name: 'Income Management', count: 30, focus: ['add income', 'edit income', 'delete income', 'category totals', 'recurring income'] },
  { name: 'Expense Management', count: 30, focus: ['add expense', 'edit expense', 'delete expense', 'category limits', 'receipt attachment'] },
  { name: 'Transaction History', count: 25, focus: ['transaction list', 'details view', 'date ranges', 'status states', 'pagination'] },
  { name: 'Budget Management', count: 20, focus: ['create budget', 'update budget', 'threshold warning', 'progress bar', 'budget reset'] },
  { name: 'Savings Goals', count: 15, focus: ['create goal', 'update contribution', 'goal progress', 'goal completion'] },
  { name: 'Reports & Analytics', count: 15, focus: ['charts', 'monthly trends', 'category breakdowns', 'insights'] },
  { name: 'Notifications & Reminders', count: 15, focus: ['budget alert', 'bill reminder', 'dismiss alert', 'unread state'] },
  { name: 'Search, Filter & Sorting', count: 20, focus: ['keyword search', 'filters', 'sort order', 'empty results'] },
  { name: 'User Profile', count: 15, focus: ['profile update', 'avatar change', 'phone update', 'account details'] },
  { name: 'Settings', count: 15, focus: ['currency setting', 'theme setting', 'privacy preferences', 'data export'] },
  { name: 'AI Budget Suggestions', count: 15, focus: ['suggestion generation', 'accept suggestion', 'dismiss suggestion', 'confidence label'] },
  { name: 'Security & Validation', count: 15, focus: ['protected routes', 'token expiry', 'input sanitization', 'biometric lock'] },
  { name: 'Offline Functionality', count: 10, focus: ['cached dashboard', 'queued transaction', 'sync recovery', 'offline banner'] },
  { name: 'Device Permissions', count: 10, focus: ['camera permission', 'storage permission', 'notification permission', 'biometric permission'] },
  { name: 'UI/UX & Responsive Testing', count: 10, focus: ['small phone layout', 'large phone layout', 'tablet layout', 'visual states'] },
  { name: 'Navigation & Gesture Testing', count: 10, focus: ['bottom navigation', 'back navigation', 'swipe gesture', 'deep link'] },
  { name: 'Performance & Stability', count: 10, focus: ['launch time', 'screen response', 'memory usage', 'long session stability'] },
  { name: 'Smoke & Regression Tests', count: 15, focus: ['critical login path', 'income smoke', 'expense smoke', 'report smoke', 'regression pass'] }
];

const typeCycle = ['Functional', 'UI', 'Validation', 'Security', 'Performance', 'End-to-End'];
const priorityCycle = ['High', 'Medium', 'Low'];
const deviceCycle = ['Pixel 7 API 35', 'Pixel 6 API 34', 'Samsung Galaxy S23', 'Android Tablet API 35'];

function slug(value) {
  return value
    .toLowerCase()
    .replace(/&/g, 'and')
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-|-$/g, '');
}

function buildCatalog() {
  let globalIndex = 1;

  return modulePlan.flatMap((module) => {
    return Array.from({ length: module.count }, (_, index) => {
      const focus = module.focus[index % module.focus.length];
      const padded = String(globalIndex).padStart(3, '0');
      const test = {
        id: `VF-${padded}`,
        module: module.name,
        title: `${module.name} - verifies ${focus} scenario ${index + 1}`,
        type: typeCycle[(globalIndex - 1) % typeCycle.length],
        priority: priorityCycle[(globalIndex - 1) % priorityCycle.length],
        platform: 'Android',
        device: deviceCycle[(globalIndex - 1) % deviceCycle.length],
        selectorHint: `[data-testid="${slug(module.name)}-${index + 1}"]`
      };

      globalIndex += 1;
      return test;
    });
  });
}

module.exports = {
  modulePlan,
  buildCatalog
};
